Feature/conn

DavidMinarsch opened the pull request:
## Proposed changes

Added the following:

- HTTP connection and config: /packages/packages/fetchai/*
- Pet Store YAML file example: /examples/http_ex/*

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

The HTTP connection.py is still largely a WIP with only the baseline functions.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        address: Address,
+        host: str = "",
+        port: int = 10000,
+        api_path: str = None,  # Directory path of the API YAML file.
```
I think you'd want it to be `api_specification_path: Optional[str]`

Kevin-Chen0 replied:
I changed it to `api_spec_path: Optional[str]`, will be in next commit.

Kevin-Chen0 replied:
For HTTPConnection, I changed it back so that it gives the user flexibility not to pass in anything into `api_spec_path` (although that may not be a good idea in terms of verification: https://stackoverflow.com/questions/51848288/in-python-should-we-set-optionalstr-none-in-the-argument-list 

But still good for demo purposes to showcase the differences between passing in the YAML file versus not.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        """
+        Initialize a connection to an RESTful API.
+
+        :param api_spec_path: the directory path of the API specification YAML source file.
```
docstrings not matching the signature

Kevin-Chen0 replied:
left it as api_spec_path

DavidMinarsch replied:
Still, your entire docstring does not match the signature.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+                )
+                return
+        else:
+            api_spec = api_path
```
All of this should be moved to the channel, just forward the `api_specification_path`

Kevin-Chen0 replied:
I have moved the following to HTTPChannel.

Kevin-Chen0 replied:
Update, it is now in _process_initial_api, but there is a check HTTPChannel that `api_spec_path` is not null.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+            raise ConnectionError(
+                "Connection not established yet. Please use 'connect()'."
+            )  # pragma: no cover
+        # self.channel.send(envelope)
```
Note, this send function is called by the multiplexer to `send` outgoing envelopes via the connection.

Kevin-Chen0 replied:
I have included send function in HTTPChannel.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        """
+        host = cast(str, connection_configuration.config.get("host"))
+        port = cast(int, connection_configuration.config.get("port"))
+        api = cast(str, connection_configuration.config.get("api")) if not "" else None
```
should be
```suggestion
        api_specification_path = cast(str, connection_configuration.config.get("api_specification_path"), "")
```

DavidMinarsch replied:
the `get` lets you specify a default value

Kevin-Chen0 replied:
fixed as `api_spec_path = cast(str, connection_configuration.config.get("api_spec_path"), "")`, will be there next commit.

Kevin-Chen0 replied:
Update, revert back to `api_spec_path = cast(str, connection_configuration.config.get("api_spec_path"))` as mypy has been given me the following errors:

```
connection.py:400: error: 'cast' expects 2 arguments
connection.py:400: error: No overload variant of "cast" matches argument types "Any", "Any", "str"
connection.py:400: note: Possible overload variants:
connection.py:400: note:     def [_T] cast(tp: Type[_T], obj: Any) -> _T
connection.py:400: note:     def cast(tp: str, obj: Any) -> Any
```

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        # self.channel.send(envelope)
+
+    async def receive(self, *args, **kwargs) -> Optional["Envelope"]:
+        """
```
Similarly, the receive method is called by the multiplexer to `receive` incoming envelopes via the connection. (So in the context of your `http` connection these are the envelopes containing the `requests`).

Kevin-Chen0 replied:
Did not change this `receive` function yet as the requests are already being picked up by do_GET, do_POST methods of HTTPHandler. Not sure whether this `receive` function will need any additional modification.

DavidMinarsch replied:
You are right, it does not need modification at this point afaikt. Just explaining to you what the supposed behaviour is.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        port: int,
+        api_spec: str,
+        connection_id: PublicId,
+        excluded_protocols: Optional[Set[PublicId]] = None,
```
you don't need the excluded protocols here. It is just needed on the `connection` object

Kevin-Chen0 replied:
I have removed from HTTPChannel, in next commit.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        self.connection_id = connection_id
+        self.in_queue = None  # type: Optional[asyncio.Queue]
+        self.loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.excluded_protocols = excluded_protocols
```
not needed

Kevin-Chen0 replied:
it has been removed

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
+            self._channel._process_request("GET", url, param)
```
Note, since this is a get request, you'd need to block here and wait for a response from this call.

So something like `response = self._channel._process_request("GET", url, param)` and then serve the response: https://pymotw.com/2/BaseHTTPServer/

Kevin-Chen0 replied:
I renamed it to `_process_incoming_request`. I first changed this function so that it returns an Envelope. Next I input this Envelope into verification helper object **APIVerificationHelper** (see below) that will then output the response.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+            content_length = int(self.headers["Content-Length"])
+            body = self.rfile.read(content_length).decode()
+
+            self._channel._process_request("POST", url, param, body)
```
same as for get request, I'd wait for a response to determine the status code and then serve the status code https://pymotw.com/2/BaseHTTPServer/

So `response = self._channel._process_request("POST", url, param, body)` and then serve it.

Kevin-Chen0 replied:
Same thing as GET. Although now both calls the function `_send` in **HTTPHandlerFactory** that takes care of the Envelope and response so that `do_GET` and `do_POST` don't have to individually.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+                # self.thread.join()
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
```
Is the factory pattern needed here? Also, PascalCase is never used for functions afaik.

Kevin-Chen0 replied:
Yes it is needed, otherwise, I would not be able to pass in the channel arg into the handler when initializing HTTPServer: `self.httpd = HTTPServer((self.host, self.port), HTTPHandlerFactory(self))`. I tried without factory but failed because HTTPHandler also need to inherit from BaseHTTPRequestHandler.

I am following the  **MakeHandlerClassFromArgv** example here, including the naming convention:

https://stackoverflow.com/questions/21631799/how-can-i-pass-parameters-to-a-requesthandler

DavidMinarsch replied:
ok

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+                    )
+                    # self.httpd = HTTPServer((self.host, self.port), handler)
+                    logger.info(f"HTTP Server has connected to port {self.port}.")
+                    self.httpd.serve_forever()
```
Isn't this blocking here?

Kevin-Chen0 replied:
Not sure what you mean. If you mean by socket, no as long as you pick an available port? I'm currently running on localhost.

DavidMinarsch replied:
as in, the call to `connect()` will never return in this setup I believe. So your multiplexer would block here. I think you need to run this in a thread.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+            "params": param,
+            "payload": body,
+        }
+        msg_bytes = json.dumps(msg).encode()
```
This is ok for now, but we will need to create an actual protocol to represent these messages once the other issues are resolved.

Kevin-Chen0 replied:
indeed, next iteration or so

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+        # Send the Envelope to the Agent's InBox.
+        assert self.in_queue is not None, "Input queue not initialized."
+        assert self.loop is not None, "Loop not initialized."
+        asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
```
You need to get the correct response envelope here as well, unpack it and return its payload etc.

Kevin-Chen0 replied:
Will try that, I'll probably have questions depending on the implementation structure.

Kevin-Chen0 replied:
Update, I currently have the **APIVerificationHelper** (see below) that directly returns the response as it is. Maybe next iteration it will return the response in an Envelope format.

DavidMinarsch replied:
As discussed offline, the APIVerificationHelper does not deal with the envelopes.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+            except Exception as e:
+                logger.error(f"Error when processing a path. Message: {str(e)}")
+
+    def _decode_path(self, path_name: str, path_value: dict, separator: str = "/"):
```
There is no point to this. Unnecessary overhead as you can just feed the specification file to the skill directly. I'd remove this method.

Kevin-Chen0 replied:
Deleted, the skill will just receive the path of the specification file.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+            self._process_api(api_spec)
+        logger.info(f"Initialized the HTTP Server on HOST: {host} and PORT: {port}")
+
+    def _process_api(self, api_spec):
```
Here I'd consume the api_specification_path as an argument and construct a helper object which lets me verify any given path

Kevin-Chen0 replied:
To start, I'll create a **PetStoreHelper** class within connection.py that emulates an "echo" handler. It takes an init Envelope from _process_api() that specifies what are the valid path and stores in an internal dict. It then uses that to verify incoming requests and returns the appropriate response.

Kevin-Chen0 replied:
I hope it is fine **APIVerificationHelper** class it nested under HTTPConnection. Next step, it may probably be its own Skill handler that receives and responds both in Envelope format through the Multiplexer.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,384 @@
+    ):
+        """Process incoming API request by packaging into Envelope and sending it in-queue.
+
+        """
```
the first thing I'd do is take the path, and feed it to the helper object for verification. All that does is it checks if the path satisfy the api specification. For now you can stub it by returning true always.

Kevin-Chen0 replied:
See my comment regarding the **APIHelper** class above.

DavidMinarsch commented on examples/http_ex/README.md:
```diff
@@ -0,0 +1,9 @@
+# HTTP agent example
```
Please move to `docs/http_connection_and_skill.md`

Kevin-Chen0 replied:
moved to `docs/http-connection-and-skill.md`

DavidMinarsch commented on packages/fetchai/connections/http/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
```
should be
```suggestion
#   Copyright 2018-2020 Fetch.AI Limited
```

DavidMinarsch replied:
same for following files

Kevin-Chen0 replied:
fixed locally, will be in next commit

DavidMinarsch commented on tests/test_cli_gui/test_search.py:
```diff
@@ -139,12 +139,18 @@ def test_real_search():
+        data[i]["description"]
+        == "The HTTP connection that wraps a web-based, RESTful API specification."
+    )
+    i += 1
```
You have duplicated the same block making your tests fail. Removing this one and keeping the one below should make the tests pass.

Kevin-Chen0 replied:
that is fixed, thanks

DavidMinarsch commented:
Description:

> HTTP connection: It accepts HTTP requests for the registered swagger api. It turns requests into Envelopes (to: agent_address, from: uuid, protocol_id: fetchai/http:0.1.0, uri: path, message: body). The message consists of the following fields (performative: HTTP verb, payload: HTTP payload, params: HTTP parameters, path: path). The message is implemented as one of our protocols. Envelopes are enqueued with the agent. A response is served once either a) a response envelope arrives from the agent or b) a timeout is hit. If the incoming response does not match the swagger api spec then it is rejected and the appropriate HTTP status code is returned.

Kevin-Chen0 commented:
**Updates:**

1) Added **APIVerificationHelper** helper object to HTTPConnection.
2) Addressed the comments on this PR.

**Next steps:**

1) Migrate the **APIVerificationHelper** out of HTTPConnection to become a HTTP Skill handler?
2) Start working on the HTTP protocol.
3) Fix any bugs arising from running AEA using the HTTPConnection, including:
`
  File ".../agents-aea/aea/decision_maker/base.py", line 529, in stop
    self._thread.join()
AttributeError: 'NoneType' object has no attribute 'join'
`

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,416 @@
+        if e is not None:
+            self.valid_paths = json.loads(e.message.decode())
+
+    def _verify(self, e: Envelope) -> str:
```
this should not consume an envelope, instead it should consume the url, path, params

Kevin-Chen0 replied:
that is fixed, will be in next commit.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,416 @@
+        method_key = msg["performative"].lower()
+
+        try:
+            status_code = "200"
```
constants should be declared at the top of the file.

Kevin-Chen0 replied:
done, will be in next commit

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,416 @@
+
+
+class APIVerificationHelper:
+    def __init__(self, e: Optional[Envelope]):
```
this should consume the api_specification_path, not the envelopes

Kevin-Chen0 replied:
fixed, also removed function _process_initial_api() as it is not needed anymore.

DavidMinarsch commented on packages/fetchai/connections/http/connection.py:
```diff
@@ -0,0 +1,416 @@
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
```
at this point I'd have something like:
```
is_valid_request, response_code = verifier.verify("GET", url, param)
if is_valid_request:
    # create envelope
    # send envelope into agent
    # wait for response envelope
   if # response envelope arrives within timeout window:
      # serve its payload
   else:
    response_code = 408 # request timeout
    self.send_response(response_code)
    self.send_header("Content-type", "text/html")
    self.end_headers()
else:
    self.send_response(response_code)
    self.send_header("Content-type", "text/html")
    self.end_headers()
```

Kevin-Chen0 replied:
Fixed, I'm still using the function `_send()`, which I renamed to `_process()` in HTTPHandler.

DavidMinarsch commented:
> **Updates:**
> 
> 1. Added **APIVerificationHelper** helper object to HTTPConnection.
> 2. Addressed the comments on this PR.
> 
> **Next steps:**
> 
> 1. Migrate the **APIVerificationHelper** out of HTTPConnection to become an HTTP Skill handler? Edit*: It will be left in HTTPConnection.
> 2. Start working on the HTTP protocol.
> 3. Fix any bugs arising from running AEA using the HTTPConnection, including:
>    `  File ".../agents-aea/aea/decision_maker/base.py", line 529, in stop self._thread.join() AttributeError: 'NoneType' object has no attribute 'join'`

1. is not a relevant step at this point. Please see my comments and the discussion we had regarding this.

Kevin-Chen0 commented:
**Updates:**

1) Address the PR comments regarding APIVerificationHelper.
2) Written the send and receive Envelope functions for HTTP Channel.

**Next steps:**

1) Write basic pet store skill/handler that returns the response Envelope.
2) Finalize the receive function in HTTP channel.

DavidMinarsch commented:
Closing this and reopening to simplify code reviews.
